scripts/linters/pylint_extensions.py

```diff
@@ -1653,100 +1653,124 @@ class SingleLineCommentChecker(checkers.BaseChecker):
+            word in line.split()[1] for word in EXCLUDED_PHRASES)
+        if (re.search(br'^# [a-z][A-Za-z]*.*$', line) and not (
+                excluded_phrase_is_present_at_beginning or
+                underscore_is_present or allowed_prefix_is_present)):
+            self.add_message(
+                'no-capital-letter-at-beginning', line=line_num)
 
-            # Single multi-line comment, ignore it.
-            if line.count(multi_line_indicator) == 2:
-                continue
+    def _check_punctuation(self, line, line_num):
+        """Checks if the comment starts with a correct punctuation.
 
-            # Flip multi-line boolean depending on whether or not we see
-            # the multi-line indicator. Possible for multiline comment to
-            # be somewhere other than the start of a line (e.g. func arg),
-            # so we can't look at start of or end of a line, which is why
-            # the case where two indicators in a single line is handled
-            # separately (i.e. one line comment with multi-line strings).
-            if multi_line_indicator in line:
-                in_multi_line_comment = not in_multi_line_comment
-
-            # Ignore anything inside a multiline comment.
-            if in_multi_line_comment:
-                continue
+        Args:
+            line: str. The current line of comment.
+            line_num: int. Line number of the current comment.
+        """
+        excluded_phrase_is_present_at_end = any(
+            word in line for word in EXCLUDED_PHRASES)
+        # Comments must end with the proper punctuation.
+        last_char_is_invalid = line[-1] not in (
+            ALLOWED_TERMINATING_PUNCTUATIONS)
+
+        excluded_phrase_at_beginning_of_line = any(
+            line[1:].startswith(word) for word in EXCLUDED_PHRASES)
+        if (last_char_is_invalid and not (
+                excluded_phrase_is_present_at_end or
+                excluded_phrase_at_beginning_of_line)):
+            self.add_message('invalid-punctuation-used', line=line_num)
 
-            next_line = ''
-            previous_line = ''
-            if line_num + 1 < file_length:
-                next_line = file_content[line_num + 1].strip()
-            if line_num > 0:
-                previous_line = file_content[line_num - 1].strip()
+    def process_tokens(self, tokens):
+        """Custom pylint checker to ensure that comments follow correct style.
 
-            # Ignore lines which are not comments.
-            if not line.startswith(b'#'):
-                continue
+        Args:
+            tokens: Token. Object to access all tokens of a module.
+        """
+        prev_line_num = -1
+        comments = []
+        comments_index = -1
 
-            # Check if comment contains any excluded phrase.
-            word_is_present_in_excluded_phrases = any(
-                word in line for word in EXCLUDED_PHRASES)
+        for (token_type, _, (line_num, _), _, line) in tokens:
```

Comment:
```suggestion
        for token_info in tokens:
```

Maybe use this and append tokens in the list*

Reply:
tokens is a tuple here, We still have to unpack tokens at a later stage that too multiple times so I find it to unpack these here.